[Feat] #24 - 일정 추가 all-day 자동 체크 로직 구현

rafa-e1 opened the pull request:
# 🔥 Pull requests

## 👷 작업한 내용
<!-- 작업한 내용을 적어주세요. -->
- 시작과 종료 시간이 24시간 이상 차이 날 경우, All-day가 자동으로 체크되도록 구현함
- 종료일이 시작일보다 정확히 1일 뒤일 경우에만 사용자가 All-day 체크박스를 해제할 수 있도록 제한함
  - 이때 시간 차이가 24시간이 넘지 않으면 수동 해제를 허용함
  - 예: 4월 1일 오후 5시 ~ 4월 2일 오후 4시 → 1일 차이지만 24시간 미만 → 해제 가능
- 사용자가 1일 차이 내에서 All-day를 해제하더라도, 해당 일정이 24시간 이상일 경우 서버에는 자동으로 `isAllDay: true`로 전송되도록 처리함
- 2일 이상 차이 날 경우에는 All-day가 자동 활성화되고, 해제할 수 없도록 고정
- 일정 시간이 24시간 미만일 경우, All-day 버튼을 비활성화(disabled) 처리함

## 📸 스크린샷
|    구현 내용    |   15 pro   |
| :-------------: | :----------: | 
| 24시간 미만일 때 All-day 버튼 비활성화 | <img src = "https://github.com/user-attachments/assets/94cd5a64-ad0e-48ce-9d0f-06323a60e75d" width ="250"> |

|    구현 내용    |   15 pro   |
| :-------------: | :----------: | 
| 1일 차이일 때만 All-day 수동 해제 가능 | <img src = "https://github.com/user-attachments/assets/3f990c17-caba-449a-87e0-9e337eb1f980" width ="250"> |

|    구현 내용    |   15 pro   |
| :-------------: | :----------: | 
| 24시간 초과 시 All-day 해제 불가 | <img src = "https://github.com/user-attachments/assets/df0ea073-b49c-47f5-b168-8bb56c0ee867" width ="250"> |

## 🖥️ 주요 코드 설명
<!-- 주요 코드에 대한 설명을 작성해주세요. -->
### `AddScheduleViewModel`
1. All-day 상태 자동 판별 로직
- `dayDifference >= 2`: 무조건 All-day
- `interval >= 86,400` (24시간): 자동 All-day
- `isAllDayManuallyToggled`가 `true`이면 사용자 의도 존중
```swift
private func checkIfAllDayShouldBeEnabled() {
    let dayDifference = numberOfDayDifference()
    let start = combine(date: startDate, time: startTime)
    let end = combine(date: endDate, time: endTime)
    let interval = end.timeIntervalSince(start)
    let shouldBeAllDay: Bool

    switch dayDifference {
    case 2...:
        shouldBeAllDay = true
    case 1, 0:
        if isAllDayManuallyToggled { return } // 수동 제어가 있는 경우 자동 설정 차단
        shouldBeAllDay = interval >= .secondsInADay
    default:
        shouldBeAllDay = false
    }

    if isAllDay != shouldBeAllDay {
        isAllDay = shouldBeAllDay
    }
}
```

2. 사용자 수동 토글 제어
- 사용자가 All-day를 수동으로 켜고 끌 수 있는 범위를 제한
- 날짜 차이가 2일 이상이면 수동으로 끌 수 없도록 제한
```swift
func toggleAllDay() {
    guard isManualAllDayChangeAllowed else {
        print("DEBUG: 2일 이상 일정은 올데이 해제 불가")
        return
    }

    isAllDay.toggle()
    isAllDayManuallyToggled = true
}
```
```swift
private var isManualAllDayChangeAllowed: Bool {
    numberOfDayDifference() < 2
}
```

3. 날짜와 시간 병합 유틸
- DatePicker에서 선택한 **날짜(date)**와 **시간(time)**을 하나의 `Date` 객체로 병합
- 이 병합된 `Date`를 기준으로 시간 간격 비교, 포맷 처리 등 수행
```swift
private func combine(date: Date, time: Date) -> Date {
    let calendar = Calendar.current
    var dateComponents = calendar.dateComponents([.year, .month, .day], from: date)
    let timeComponents = calendar.dateComponents([.hour, .minute, .second], from: time)

    dateComponents.hour = timeComponents.hour
    dateComponents.minute = timeComponents.minute
    dateComponents.second = timeComponents.second

    return calendar.date(from: dateComponents) ?? date
}
```

4. 24시간 이상 판단 (`isEventLongerThan24Hours`)
- 일정이 24시간 이상인지 여부를 판단하여 All-day 조건에 활용
```swift
func isEventLongerThan24Hours() -> Bool {
    let start = combine(date: startDate, time: startTime)
    let end = combine(date: endDate, time: endTime)

    return end.timeIntervalSince(start) >= .secondsInADay
}
```
```swift
extension TimeInterval {
    static let secondsInADay: TimeInterval = 86_400
}
```

5. All-day 상태에 따라 시간 값 업데이트
- All-day일 경우 시간 값을 하루 시작으로 고정
- All-day가 아닐 경우 자동으로 2시간 간격으로 라운딩된 시간으로 설정
```swift
private func updateTimesForAllDayStatus() {
    if isAllDay {
        startTime = startDate.startOfDay
        endTime = endDate.startOfDay
    } else {
        guard let (roundedStart, roundedEnd) = Self.makeRoundedStartAndEnd() else { return }
        startTime = roundedStart
        endTime = roundedEnd
    }
}
```

6. 사용자 UX 흐름 정리

| 상황 | 동작 |
| :-- | :-- |
| 날짜 차이 2일 이상 | All-day 무조건 true, 수동 해제 불가 |
| 날짜 차이 1일, 시간 24시간 이상 | 자동 All-day, 수동 해제 가능 |
| 수동 해제 이후 다시 24시간 이상 설정 | 자동으로 All-day 복구 |
| 시간 피커 | All-day일 때 비활성화 및 흐림 처리 |

## ✅ Check List
- [x] Merge 대상 브랜치가 올바른가?
- [x] 최종 코드가 에러 없이 잘 동작하는가?
- [x] 전체 변경사항이 500줄을 넘지 않는가?

## 📟 관련 이슈
- Resolved: #24

rafa-e1 committed:
[Feat] #24 - TimeInterval 상수 파일 추가

TimeInterval 확장을 통해 하루의 초를 나타내는 상수를 정의함

rafa-e1 committed:
[Refactor] #24 - 날짜 및 시간 변수명 수정

startsDate, endsDate, selectedStartTime, selectedEndTime을 각각 startDate, endDate, startTime, endTime으로 변경

rafa-e1 committed:
[Feat] #24 - 24시간 이상일 때 all-day 자동 체크 로직 구현

- isAllDay 상태 변경 시 수동 조작 여부를 확인하는 로직 추가
- 이벤트가 24시간 이상일 경우 자동으로 all-day로 설정되도록 수정
- 날짜 및 시간 결합을 위한 combine 함수 추가
- 일정 기간이 2일 이상인 경우 all-day 해제 불가하도록 제한
- API 요청 시 isAllDay 값을 적절히 설정하여 전송하도록 변경

rafa-e1 committed:
[Refactor] #24 - 변경된 변수명 반영

rafa-e1 committed:
[Feat] #24 - 'All-day' 체크박스 UI 개선

- 뷰모델의 이벤트 길이에 따라 '하루 종일' 체크박스의 색상과 투명도를 조정함
- 이벤트가 24시간 이상인 경우에만 활성화되도록 버튼을 비활성화하는 로직 추가

mcrkgus commented on Memento-iOS/Memento-iOS/Source/Presentation/Add/AddSchedule/ViewModel/AddScheduleViewModel.swift:
```diff
@@ -12,15 +12,23 @@ final class AddScheduleViewModel: ObservableObject, TagSelectable {
-    @Published var endsDate: Date
-    @Published var selectedStartTime: Date
-    @Published var selectedEndTime: Date
+    @Published var startDate: Date { didSet { checkIfAllDayShouldBeEnabled() } }
+    @Published var startTime: Date { didSet { checkIfAllDayShouldBeEnabled() } }
+    @Published var endDate: Date { didSet { checkIfAllDayShouldBeEnabled() } }
+    @Published var endTime: Date { didSet { checkIfAllDayShouldBeEnabled() } }
```
p3
Good 깔끔해요

mcrkgus commented on Memento-iOS/Memento-iOS/Source/Presentation/Add/AddSchedule/ViewModel/AddScheduleViewModel.swift:
```diff
@@ -52,36 +63,43 @@ final class AddScheduleViewModel: ObservableObject, TagSelectable {
-    
-    func postMakeScheduleNotiFication() {
+
+    private func postMakeScheduleNotiFication() {
```
p3
private 메소드들은 따로 정리해도 좋을 것 같아요! 
```swift
private extension {
 func ~~~~ 
}
```

rafa-e1 replied:
넵 감사합니다

mcrkgus commented on Memento-iOS/Memento-iOS/Source/Presentation/Add/AddSchedule/ViewModel/AddScheduleViewModel.swift:
```diff
@@ -147,39 +165,84 @@ final class AddScheduleViewModel: ObservableObject, TagSelectable {
+    // MARK: - User Interaction
+
     func toggleAllDay() {
+        guard isManualAllDayChangeAllowed else {
```
p3
오.. guard문을 직접 사용한 적은 없었는데 좋네요!!! 굳굳

mcrkgus reviewed (approved):
고생하셨습니당~!~!

sem-git reviewed (approved):
로직 구현하느라 고생 많으셨습니다 ~!

rafa-e1 committed:
[Refactor] #24 - AddScheduleViewModel의 private 메서드 분리 및 코드 정리

Private extension을 사용하여 AddScheduleViewModel의 private 메서드를 그룹화하고, 불필요한 private 접근 제어자를 제거함. 각 기능별로 MARK 주석을 추가하여 가독성을 높였음.
